### ISSUE TITLE
Feedback for pipeline

### ISSUE THREAD

**Issue by pstokespmb** [popgenmethods/smcpp#147]
I am performing SMC++ on a much larger dataset than previously. I was hoping to get some input about my general pipeline:

1.) Use paleomix for the mapping (uses BWA)
2.) Pass BAM files into ANGSD (capturing variant and invariant sites), masking repeat region sites (hard to map) --> VCF with genotype likelihoods
2 - question.) Is it important to capture invariant sites (sites where ALL individuals are homozygous ref)? This blew up my file sizes a lot
3.) Create masking BED file from known reference repeat regions + sites not called as variant or invariant (due to not hitting quality thresholds or maybe not sequenced by chance or low coverage)
4.) Pass VCF and BED into SMC++

Does this pipeline seem okay to you?

My output now differs from previous output, but there are also now significantly more samples, variant calling was done differently, and last time I did not capture invariant sites.

**Comment by terhorst** [comment]
I think that sound reasonable. You should not need to embed all homozygous sites as long as you used the `--mask` option to vcf2smc with your BED files. If you are unsure about this, the SMC file format is gzipped plain text, so you can look at it and verify that it correctly represents your data. The format is documented here: https://github.com/popgenmethods/smcpp#id25.

**Comment by pstokespmb** [comment]
A bit ago when I was using fewer samples we were able to generate a plot that agreed with our hypothesis as well as previous projects' results. During this time, we kept only sites in the VCF that had at least one sample with variation from the reference. 

This time, with way more samples (10x more), we generated the VCF to retain sites with at least one individual showing variation from ref (same as above) as well as sites that are homozygous ref across all individuals (that have coverage for that site).

The plot has the same bottleneck pattern, but >250,000 generations in the past. We expect it to be ~4,000 generations in the past. 

In short, is it necessary or recommended to keep sites that are homozygous ref? Nearly nothing else is different between the two runs except the number of individuals in the VCF and the homozygous sites (I even ran the same DIs and non-DIs).

**Comment by DeondeJager** [comment]
Hi @pstokespmb,

Sorry for tagging on to your post here. I just wanted to ask: How are you converting your VCF from ANGSD with genotype likelihoods to a VCF recognised by SMC++? i.e. with genotypes coded as 0/1 for example, or ./. for missing data. SMC++ of course gives an error when I just pass in the VCF with genotype likelihoods, saying it expected a diploid genotype but found nothing.

**Comment by pstokespmb** [comment]
Hey @DeondeJager,

ANGSD will output a VCF file with "-doVCF 1". The VCF file does display predicted genotypes based on genotype likelihoods. I have had no problems using SMC++ on these VCF files because SMC++ only uses the genotypes and ignores all FORMAT information. **What version of ANGSD are you using?** Recently, I needed to update the version of ANGSD on the cluster because it was not outputting a VCF with genotypes associated with genotype likelihoods, but rather was just outputting the genotype likelihoods. 

It sounds to me that there might be a problem position in your VCF. In my experience, SMC++ will indicate what position an error was thrown. I would look more closely at that position.

**Comment by DeondeJager** [comment]
Thanks for the reply and info! I was indeed using the `-doVcf 1` option, but with an older version of ANGSD: v0.918. Which version did you update to? I'll update to the same version on my side and see if it solves my problem.

SMC++ indicates an error with the first position in the VCF file as follows: 
"RuntimeError: Expected a diploid genotype at position 75 for individual KNP1 but found:
[]"

**Comment by pstokespmb** [comment]
Hey @DeondeJager,

Yep, sounds like a version thing. I am running v.0.931. The version on the cluster is v.0.919. 

Just to confirm, when you look at your VCF, there are no explicit genotypes right? ie (1/1), (0/1), or even missing (./.)

I suspect updating ANGSD and getting VCFs with genotypes will help!

Just out of curiosity, what type of data are you working with and what coverage? I recently had a dissertation committee meeting with Rasmus Nielsen (corresponding author of ANGSD) and he had some good feedback about my data. It might be beneficial to you if you are working with similar data.

**Comment by DeondeJager** [comment]
Thanks @pstokespmb,

I'll update the v0.931, do a test run a report back here whether it worked, in case others come across this thread. 

Yes, confirming that when I look at the VCF there are no explicit genotypes, just likelihoods.

I am working with low-coverage (mean depth = 7) whole-genome resequencing data (Illumina PE 150bp reads with 350bp insert sizes). 

I came across your post on the ANGSD Github page regarding invariant sites: https://github.com/ANGSD/angsd/issues/288. I see the vcf output has been deprecated, but that bcf output is possible.

**Comment by DeondeJager** [comment]
You are in a privileged position to have Rasmus Nielsen on your dissertation committee! :) Any advice or tips on my data and the analyses would be welcomed and appreciated, if it is similar to yours.

**Comment by pstokespmb** [comment]
@DeondeJager Rasmus is great! He is such a fun person to have on my team for sure!

I think taking a genotype likelihood approach is a good way to move forward with lower coverage samples. I was initially using ANGSD as well, because most of my data ranges from 3x-20x coverage, but I found that I was having too many inconsistencies with my SMC++ output. It was determined that it was because of using a poor quality VCF file (ti/tv=1.6ish). Plant genomes are notoriously crappy, so I decided to move forward with GATK + VQSR to hopefully generate a better VCF for my organism (sunflower). I suspect that ANGSD will be great for moving forward with SMC++ for your system! 

I also found that our system has a ton of repeat regions (like 70% of our 3.6Gb genome is repeat), which also contributes to poor read mapping, but hey, gotta make due with what we have :)

**Comment by DeondeJager** [comment]
@pstokespmb thanks for sharing your experience! :) As far as I can remember the buffalo genome does not have an unusual amount of repetitive regions, but I'll just double-check the original genome paper to make sure and also check my ti/tv ratio after generating my VCF file (thanks for the timely reminder)! 

Wow, yeah, that is a lot of repeats! Indeed, we make do with what we have! :) Good luck on that side; I'm sure GATK + VQSR will get you what you need!

**Comment by DeondeJager** [comment]
> I'll update the v0.931, do a test run a report back here whether it worked, in case others come across this thread.
> 
> Yes, confirming that when I look at the VCF there are no explicit genotypes, just likelihoods.

Success! I updated to the latest version of ANGSD, which is v0.933 (via git clone following the instructions at the bottom of this page: https://github.com/ANGSD/angsd). I can confirm that the VCF file now includes explicit genotypes! :) I used the `-doBcf 1` flag with this version of ANGSD and then converted the BCF to VCF using `bcftools view -Oz -o output.vcf.gz input.bcf` so I could look at the file and confirm explicit genotypes are present. SMC++ then happily accepted this VCF file. Thanks again for your help, @pstokespmb!
